### PR TITLE
REG: DataFrame.shift with axis=1 and CategoricalIndex columns

### LINKED ISSUE #38434: REGR: DataFrame.shift(axis=1) raises TypeError when columns is CategoricalIndex
A regression since 1.1:

```python
>>> ci = pd.CategoricalIndex(["a", "b"])
>>> df = pd.DataFrame([[1, 2], [3, 4]], index=ci, columns=ci)
>>> df.shift(axis=0)  # ok
     a    b
a  NaN  NaN
b  1.0  2.0
>>> df.shift(axis=1)  # not ok, worked in v1.1
TypeError: 'fill_value=0' is not present in this Categorical's categories
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #38434
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by topper-123** [comment]
Thanks for taking this.

This PR should IMO be part of 1.2, since it’s a regression since 1.1.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4584,20 +4584,23 @@ def shift(
...
+            label = self.columns[0]
+
             if periods > 0:
                 result = self.iloc[:, :-periods]
                 for col in range(min(ncols, abs(periods))):
                     # TODO(EA2D): doing this in a loop unnecessary with 2D EAs
                     # Define filler inside loop so we get a copy
                     filler = self.iloc[:, 0].shift(len(self))
-                    result.insert(0, col, filler, allow_duplicates=True)
+                    result.insert(0, label, filler, allow_duplicates=True)
```
Comment (topper-123) [inline pandas/core/frame.py:4596]: Does this work correctly for periods > 1? Looks like it inserts the same label in all locations, where Id think it should insert `self.columns[col]`?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4584,20 +4584,23 @@ def shift(
...
             else:
                 result = self.iloc[:, -periods:]
                 for col in range(min(ncols, abs(periods))):
                     # Define filler inside loop so we get a copy
                     filler = self.iloc[:, -1].shift(len(self))
                     result.insert(
-                        len(result.columns), col, filler, allow_duplicates=True
+                        len(result.columns), label, filler, allow_duplicates=True
```
Comment (topper-123) [inline pandas/core/frame.py:4603]: Same comment as above.

File: pandas/tests/frame/methods/test_shift.py [hunk 3]
```diff
@@ -292,3 +292,12 @@ def test_shift_dt64values_int_fill_deprecated(self):
 
         expected = DataFrame({"A": [pd.Timestamp(0), pd.Timestamp(0)], "B": df2["A"]})
         tm.assert_frame_equal(result, expected)
+
+    def test_shift_axis1_categorical_columns(self):
+        # GH#38434
+        ci = CategoricalIndex(["a", "b"])
+        df = DataFrame([[1, 2], [3, 4]], index=ci, columns=ci)
+        result = df.shift(axis=1)
```
Comment (topper-123) [inline pandas/tests/frame/methods/test_shift.py:300]: Can you test a 3-column dataframe with `periods=2` also?

**Review by topper-123 — CHANGES_REQUESTED** [review]

**Comment by simonjayhawkins** [comment]
> This PR should IMO be part of 1.2, since it’s a regression since 1.1.

and whatsnew probably not necessary

File: pandas/core/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:4596]: This should be benign because we set result.columns directly below.  ill add a test to be on the safe side

**Comment by jbrockmendel** [comment]
added test case with periods=2, removed whatsnew

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Pull request merged**